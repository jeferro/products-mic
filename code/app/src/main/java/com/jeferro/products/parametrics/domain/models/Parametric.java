package com.jeferro.products.parametrics.domain.models;

import com.jeferro.products.parametrics.domain.exceptions.ParametricValueNotFoundException;
import com.jeferro.products.parametrics.domain.models.values.ParametricValueId;
import com.jeferro.products.parametrics.domain.models.values.ParametricValues;
import com.jeferro.shared.ddd.domain.models.projection.Projection;
import com.jeferro.shared.ddd.domain.utils.ValueValidationUtils;
import com.jeferro.shared.locale.domain.models.LocalizedField;
import lombok.Getter;

@Getter
public class Parametric extends Projection<ParametricId> {

    private final LocalizedField name;

    private final ParametricValues values;

    public Parametric(
            ParametricId id,
            LocalizedField name,
            ParametricValues values) {
        super(id);

        this.name = name;
        this.values = values;
    }

    public static Parametric createOf(ParametricId id, LocalizedField name, ParametricValues values) {
        ValueValidationUtils.isNotNull(name, "name", Parametric.class);
        ValueValidationUtils.isNotEmpty(values, "values", Parametric.class);

        return new Parametric(id, name, values);
    }

    public boolean validate(ParametricValueId parametricValueId) {
        if (notContainsParametricValue(parametricValueId)) {
            throw ParametricValueNotFoundException.createOf(this, parametricValueId);
        }

        return true;
    }

    private boolean notContainsParametricValue(ParametricValueId parametricValueId) {
        return values.stream()
                .noneMatch(parametricValue -> parametricValue.hasSameId(parametricValueId));
    }
}
