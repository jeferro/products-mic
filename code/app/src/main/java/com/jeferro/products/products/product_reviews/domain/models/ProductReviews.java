package com.jeferro.products.products.product_reviews.domain.models;

import com.jeferro.shared.ddd.domain.models.aggregates.EntityCollection;

import java.util.Arrays;
import java.util.List;

public class ProductReviews extends EntityCollection<ProductReviewId, ProductReview> {

    public ProductReviews() {
        super();
    }

    public ProductReviews(List<ProductReview> entities) {
        super(entities);
    }

    public static ProductReviews createOf(ProductReview... productReview) {
        var entities = Arrays.asList(productReview);

        return new ProductReviews(entities);
    }
}
