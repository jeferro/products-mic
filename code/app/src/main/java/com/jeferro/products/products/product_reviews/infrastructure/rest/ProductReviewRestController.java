package com.jeferro.products.products.product_reviews.infrastructure.rest;

import com.jeferro.products.generated.rest.v1.apis.ProductReviewsApi;
import com.jeferro.products.generated.rest.v1.dtos.CreateProductReviewInputRestDTO;
import com.jeferro.products.generated.rest.v1.dtos.ProductReviewRestDTO;
import com.jeferro.products.generated.rest.v1.dtos.UpdateProductReviewInputRestDTO;
import com.jeferro.products.products.product_reviews.infrastructure.rest.mappers.ProductReviewRestMapper;
import com.jeferro.shared.ddd.application.bus.HandlerBus;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
@RequiredArgsConstructor
public class ProductReviewRestController implements ProductReviewsApi {

    private final ProductReviewRestMapper productReviewRestMapper = ProductReviewRestMapper.INSTANCE;

    private final HandlerBus handlerBus;

    @Override
    public List<ProductReviewRestDTO> searchProductReviews(String productCode) {
        var params = productReviewRestMapper.toSearchProductsParams(productCode);

        var productReviews = handlerBus.execute(params);

        return productReviewRestMapper.toDTO(productReviews);
    }

    @Override
    public ProductReviewRestDTO createProductReview(CreateProductReviewInputRestDTO inputRestDTO) {
        var params = productReviewRestMapper.toCreateProductReviewParams(inputRestDTO);

        var productReview = handlerBus.execute(params);

        return productReviewRestMapper.toDTO(productReview);
    }

    @Override
    public ProductReviewRestDTO getProductReview(String productReviewId) {
        var params = productReviewRestMapper.toGetProductReviewParams(productReviewId);

        var productReview = handlerBus.execute(params);

        return productReviewRestMapper.toDTO(productReview);
    }

    @Override
    public ProductReviewRestDTO updateProductReview(String productReviewId,
                                                    UpdateProductReviewInputRestDTO inputRestDTO) {
        var params = productReviewRestMapper.toUpdateProductReviewParams(productReviewId, inputRestDTO);

        var productReview = handlerBus.execute(params);

        return productReviewRestMapper.toDTO(productReview);
    }

    @Override
    public ProductReviewRestDTO deleteProductReview(String productReviewId) {
        var params = productReviewRestMapper.toDeleteProductReviewParams(productReviewId);

        var productReview = handlerBus.execute(params);

        return productReviewRestMapper.toDTO(productReview);
    }
}
