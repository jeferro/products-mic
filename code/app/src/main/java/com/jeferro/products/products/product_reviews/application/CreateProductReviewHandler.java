package com.jeferro.products.products.product_reviews.application;

import com.jeferro.products.products.product_reviews.application.params.CreateProductReviewParams;
import com.jeferro.products.products.product_reviews.domain.exceptions.ProductReviewAlreadyExistsException;
import com.jeferro.products.products.product_reviews.domain.models.ProductReview;
import com.jeferro.products.products.product_reviews.domain.models.ProductReviewId;
import com.jeferro.products.products.product_reviews.domain.repositories.ProductReviewsRepository;
import com.jeferro.products.products.products.domain.models.ProductCode;
import com.jeferro.products.products.products.domain.repositories.ProductsRepository;
import com.jeferro.shared.ddd.application.Handler;
import com.jeferro.shared.ddd.domain.events.EventBus;
import com.jeferro.shared.ddd.domain.exceptions.auth.ForbiddenException;
import com.jeferro.shared.ddd.domain.models.auth.Auth;
import com.jeferro.shared.ddd.domain.models.auth.UserAuth;
import com.jeferro.shared.ddd.domain.models.context.Context;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.util.Set;

import static com.jeferro.products.shared.application.Roles.USER;

@Component
@RequiredArgsConstructor
public class CreateProductReviewHandler extends Handler<CreateProductReviewParams, ProductReview> {

    private final ProductsRepository productsRepository;

    private final ProductReviewsRepository productReviewsRepository;

    private final EventBus eventBus;

    @Override
    public Set<String> getMandatoryUserRoles() {
        return Set.of(USER);
    }

    @Override
    public ProductReview execute(Context context, CreateProductReviewParams params) {
        Auth auth = context.getAuth();

        var productCode = ensureProductExists(params);

        ensureUserIsAuthenticated(auth);

        ensureProductReviewDoesNotExists(context, productCode);

        return createProductReview(context, params, productCode);
    }

    private void ensureUserIsAuthenticated(Auth auth) {
        if (!(auth instanceof UserAuth)) {
            throw ForbiddenException.createOfNotUserAuth(auth);
        }
    }

    private ProductCode ensureProductExists(CreateProductReviewParams params) {
        var productCode = params.getProductCode();

        var product = productsRepository.findByIdOrError(productCode);

        return product.getCode();
    }

    private void ensureProductReviewDoesNotExists(Context context, ProductCode productCode) {
        var auth = context.getAuth();

        var productReviewId = ProductReviewId.createOf(auth, productCode);

        productReviewsRepository.findById(productReviewId)
                .ifPresent(current -> {
                    throw ProductReviewAlreadyExistsException.createOf(productReviewId);
                });
    }

    private ProductReview createProductReview(Context context, CreateProductReviewParams params, ProductCode productCode) {
        var auth = context.getAuth();
        var locale = context.getLocale();

        var comment = params.getComment();

        var productReview = ProductReview.createOf(productCode, locale, comment, auth);

        productReviewsRepository.save(productReview);

        eventBus.sendAll(productReview);
        return productReview;
    }
}
