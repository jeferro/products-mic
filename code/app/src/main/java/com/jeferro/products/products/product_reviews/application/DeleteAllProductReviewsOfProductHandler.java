package com.jeferro.products.products.product_reviews.application;

import com.jeferro.products.products.product_reviews.application.params.DeleteAllProductReviewsOfProductParams;
import com.jeferro.products.products.product_reviews.domain.models.ProductReview;
import com.jeferro.products.products.product_reviews.domain.repositories.ProductReviewsRepository;
import com.jeferro.shared.ddd.application.Handler;
import com.jeferro.shared.ddd.domain.events.EventBus;
import com.jeferro.shared.ddd.domain.models.context.Context;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.util.Set;

import static com.jeferro.products.shared.application.Roles.ADMIN;

@Component
@RequiredArgsConstructor
public class DeleteAllProductReviewsOfProductHandler extends Handler<DeleteAllProductReviewsOfProductParams, Void> {

    private final ProductReviewsRepository productReviewsRepository;

    private final EventBus eventBus;

    @Override
    public Set<String> getMandatoryUserRoles() {
        return Set.of(ADMIN);
    }

    @Override
    public Void execute(Context context, DeleteAllProductReviewsOfProductParams params) {
        var productCode = params.getProductCode();

        var productReviews = productReviewsRepository.findAllByProductCode(productCode);

        if (productReviews.isEmpty()) {
            return null;
        }

        productReviews.forEach(ProductReview::deleteBySystem);

        var productReviewIds = productReviews.getIds();

        productReviewsRepository.deleteAllById(productReviewIds);

        eventBus.sendAll(productReviews);

        return null;
    }
}
