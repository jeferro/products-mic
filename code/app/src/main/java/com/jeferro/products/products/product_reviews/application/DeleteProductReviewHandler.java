package com.jeferro.products.products.product_reviews.application;

import com.jeferro.products.products.product_reviews.application.params.DeleteProductReviewParams;
import com.jeferro.products.products.product_reviews.domain.models.ProductReview;
import com.jeferro.products.products.product_reviews.domain.repositories.ProductReviewsRepository;
import com.jeferro.shared.ddd.application.Handler;
import com.jeferro.shared.ddd.domain.events.EventBus;
import com.jeferro.shared.ddd.domain.models.context.Context;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.util.Set;

import static com.jeferro.products.shared.application.Roles.USER;

@Component
@RequiredArgsConstructor
public class DeleteProductReviewHandler extends Handler<DeleteProductReviewParams, ProductReview> {

    private final ProductReviewsRepository productReviewsRepository;

    private final EventBus eventBus;

    @Override
    public Set<String> getMandatoryUserRoles() {
        return Set.of(USER);
    }

    @Override
    public ProductReview execute(Context context, DeleteProductReviewParams params) {
        var productReview = ensureProductReviewExists(params);

        return deleteProductReview(context, productReview);
    }

    private ProductReview ensureProductReviewExists(DeleteProductReviewParams params) {
        var productReviewId = params.getProductReviewId();

        return productReviewsRepository.findByIdOrError(productReviewId);
    }

    private ProductReview deleteProductReview(Context context, ProductReview productReview) {
        var auth = context.getAuth();

        productReview.deleteByUser(auth);

        productReviewsRepository.deleteById(productReview.getId());

        eventBus.sendAll(productReview);

        return productReview;
    }
}
