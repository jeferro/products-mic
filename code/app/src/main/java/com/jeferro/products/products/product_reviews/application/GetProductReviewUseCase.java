package com.jeferro.products.products.product_reviews.application;

import com.jeferro.products.products.product_reviews.application.params.GetProductReviewParams;
import com.jeferro.products.products.product_reviews.domain.models.ProductReview;
import com.jeferro.products.products.product_reviews.domain.repositories.ProductReviewsRepository;
import com.jeferro.shared.ddd.application.UseCase;
import com.jeferro.shared.ddd.domain.models.context.Context;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.util.Set;

import static com.jeferro.products.shared.application.Roles.USER;

@Component
@RequiredArgsConstructor
public class GetProductReviewUseCase extends UseCase<GetProductReviewParams, ProductReview> {

    private final ProductReviewsRepository productReviewsRepository;

    @Override
    public Set<String> getMandatoryUserRoles() {
        return Set.of(USER);
    }

    @Override
    public ProductReview execute(Context context, GetProductReviewParams params) {
        var productReviewId = params.getProductReviewId();

        return productReviewsRepository.findByIdOrError(productReviewId);
    }
}
