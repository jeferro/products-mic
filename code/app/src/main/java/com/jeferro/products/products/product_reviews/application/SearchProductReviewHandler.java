package com.jeferro.products.products.product_reviews.application;

import com.jeferro.products.products.product_reviews.application.params.SearchProductReviewParams;
import com.jeferro.products.products.product_reviews.domain.models.ProductReview;
import com.jeferro.products.products.product_reviews.domain.repositories.ProductReviewsRepository;
import com.jeferro.shared.ddd.application.Handler;
import com.jeferro.shared.ddd.domain.models.aggregates.PaginatedList;
import com.jeferro.shared.ddd.domain.models.context.Context;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.util.Set;

import static com.jeferro.products.shared.application.Roles.USER;

@Component
@RequiredArgsConstructor
public class SearchProductReviewHandler extends Handler<SearchProductReviewParams, PaginatedList<ProductReview>> {

    private final ProductReviewsRepository productReviewsRepository;

    @Override
    public Set<String> getMandatoryUserRoles() {
        return Set.of(USER);
    }

    @Override
    public PaginatedList<ProductReview> execute(Context context, SearchProductReviewParams params) {
        var productCode = params.getProductCode();

        return productReviewsRepository.findAllByProductCode(productCode);
    }
}
