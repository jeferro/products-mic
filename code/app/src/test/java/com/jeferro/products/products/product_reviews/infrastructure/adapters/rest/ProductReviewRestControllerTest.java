package com.jeferro.products.products.product_reviews.infrastructure.adapters.rest;

import com.jeferro.products.products.product_reviews.domain.models.ProductReviewMother;
import com.jeferro.products.products.product_reviews.domain.models.ProductReviews;
import com.jeferro.products.products.product_reviews.infrastructure.rest.ProductReviewRestController;
import com.jeferro.products.products.products.domain.models.ProductMother;
import com.jeferro.products.shared.application.StubUseCaseBus;
import com.jeferro.products.shared.infrastructure.adapters.rest.RestControllerTest;
import com.jeferro.products.shared.infrastructure.adapters.utils.ApprovalUtils;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;

@WebMvcTest(ProductReviewRestController.class)
class ProductReviewRestControllerTest extends RestControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private StubUseCaseBus stubHandlerBus;

    @Test
    void execute_list_product_reviews_on_request() throws Exception {
        var apple = ProductMother.apple();
        var productReviews = ProductReviews.createOf(
                ProductReviewMother.userReviewOfApple(),
                ProductReviewMother.adminReviewOfApple()
        );
        stubHandlerBus.init(productReviews);

        var requestBuilder = MockMvcRequestBuilders.get("/v1/product-reviews?productCode=" + apple.getId())
                .contentType(MediaType.APPLICATION_JSON)
                .header(HttpHeaders.ACCEPT_LANGUAGE, ACCEPT_LANGUAGE_EN)
                .header(HttpHeaders.AUTHORIZATION, AUTHORIZATION_USER_TOKEN);

        var response = mockMvc.perform(requestBuilder)
                .andReturn()
                .getResponse();

        ApprovalUtils.verifyAll(stubHandlerBus.getFirstParamOrError(),
                response.getStatus(),
                response.getContentAsString());
    }

    @Test
    void execute_create_product_on_request() throws Exception {
        var userReviewOfProduct = ProductReviewMother.userReviewOfApple();
        stubHandlerBus.init(userReviewOfProduct);

        var requestContent = """
                {
                  "productCode": "%s",
                  "comment": "%s"
                }"""
                .formatted(userReviewOfProduct.getProductCode(), userReviewOfProduct.getComment());

        var requestBuilder = MockMvcRequestBuilders.post("/v1/product-reviews")
                .contentType(MediaType.APPLICATION_JSON)
                .header(HttpHeaders.ACCEPT_LANGUAGE, ACCEPT_LANGUAGE_EN)
                .header(HttpHeaders.AUTHORIZATION, AUTHORIZATION_USER_TOKEN)
                .content(requestContent);

        var response = mockMvc.perform(requestBuilder)
                .andReturn()
                .getResponse();

        ApprovalUtils.verifyAll(stubHandlerBus.getFirstParamOrError(),
                response.getStatus(),
                response.getContentAsString());
    }

    @Test
    void execute_get_product_on_request() throws Exception {
        var userReviewOfProduct = ProductReviewMother.userReviewOfApple();
        stubHandlerBus.init(userReviewOfProduct);

        var requestBuilder = MockMvcRequestBuilders.get("/v1/product-reviews/" + userReviewOfProduct.getId())
                .contentType(MediaType.APPLICATION_JSON)
                .header(HttpHeaders.ACCEPT_LANGUAGE, ACCEPT_LANGUAGE_EN)
                .header(HttpHeaders.AUTHORIZATION, AUTHORIZATION_USER_TOKEN);

        var response = mockMvc.perform(requestBuilder)
                .andReturn()
                .getResponse();

        ApprovalUtils.verifyAll(stubHandlerBus.getFirstParamOrError(),
                response.getStatus(),
                response.getContentAsString());
    }

    @Test
    void execute_update_product_on_request() throws Exception {
        var userReviewOfProduct = ProductReviewMother.userReviewOfApple();
        stubHandlerBus.init(userReviewOfProduct);

        var requestContent = """
                {
                  "comment": "%s"
                }"""
                .formatted(userReviewOfProduct.getComment());

        var requestBuilder = MockMvcRequestBuilders.put("/v1/product-reviews/" + userReviewOfProduct.getId())
                .contentType(MediaType.APPLICATION_JSON)
                .header(HttpHeaders.ACCEPT_LANGUAGE, ACCEPT_LANGUAGE_EN)
                .header(HttpHeaders.AUTHORIZATION, AUTHORIZATION_USER_TOKEN)
                .content(requestContent);

        var response = mockMvc.perform(requestBuilder)
                .andReturn()
                .getResponse();

        ApprovalUtils.verifyAll(stubHandlerBus.getFirstParamOrError(),
                response.getStatus(),
                response.getContentAsString());
    }

    @Test
    void execute_delete_product_on_request() throws Exception {
        var userReviewOfProduct = ProductReviewMother.userReviewOfApple();
        stubHandlerBus.init(userReviewOfProduct);

        var requestBuilder = MockMvcRequestBuilders.delete("/v1/product-reviews/" + userReviewOfProduct.getId())
                .contentType(MediaType.APPLICATION_JSON)
                .header(HttpHeaders.ACCEPT_LANGUAGE, ACCEPT_LANGUAGE_EN)
                .header(HttpHeaders.AUTHORIZATION, AUTHORIZATION_USER_TOKEN);

        var response = mockMvc.perform(requestBuilder)
                .andReturn()
                .getResponse();

        ApprovalUtils.verifyAll(stubHandlerBus.getFirstParamOrError(),
                response.getStatus(),
                response.getContentAsString());
    }
}
