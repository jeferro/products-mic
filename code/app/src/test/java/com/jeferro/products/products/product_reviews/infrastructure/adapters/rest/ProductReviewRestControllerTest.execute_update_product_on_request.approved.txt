data[0] = UpdateProductReviewParams[comment=Comment about apple,productReviewId=user::apple]
data[1] = 200
data[2] = {"id":{"productCode":"apple","username":"user"},"locale":"en-US","comment":"Comment about apple"}
